rpc/src/rpc/native/txn.rs

```diff
@@ -0,0 +1,283 @@
+use tokio::sync::Mutex;
+use trace_decoder::trace_protocol::ContractCodeUsage;
+use trace_decoder::trace_protocol::TxnTrace;
+use trace_decoder::trace_protocol::{TxnInfo, TxnMeta};
+
+use super::tracing_options_diff;
+use super::{tracing_options, Block, Http};
+
+/// Processes the transactions in the given block and updates the code db.
+pub(super) async fn process_transactions(
+    block: &Block<H256>,
+    provider: Arc<Provider<RetryClient<Http>>>,
+    code_db: Arc<Mutex<HashMap<H256, Vec<u8>>>>,
+) -> Result<Vec<TxnInfo>> {
+    let mut futures_ordered = FuturesOrdered::new();
+
+    for tx_hash in &block.transactions {
+        let provider = provider.clone();
+        let code_db = code_db.clone();
+        futures_ordered.push_back(async move {
+            super::txn::process_transaction(provider, tx_hash, code_db).await
+        });
+    }
+
+    let tx_infos: Vec<TxnInfo> = futures_ordered.try_collect().await?;
+    Ok(tx_infos)
+}
+
+/// Processes the transaction with the given transaction hash and updates the
+/// accounts state.
+async fn process_transaction(
+    provider: Arc<Provider<RetryClient<Http>>>,
+    tx_hash: &H256,
+    code_db: Arc<Mutex<HashMap<H256, Vec<u8>>>>,
+) -> Result<TxnInfo> {
+    let (tx, tx_receipt, pre_trace, diff_trace) = fetch_tx_data(provider, tx_hash).await?;
+    let tx_meta = TxnMeta {
+        byte_code: tx.rlp().to_vec(),
+        new_txn_trie_node_byte: tx.rlp().to_vec(),
+        new_receipt_trie_node_byte: compute_receipt_bytes(&tx_receipt),
+        gas_used: tx_receipt.gas_used.unwrap().as_u64(),
+    };
+
+    let access_list = parse_access_list(tx.access_list.unwrap_or_default());
+
+    let tx_traces = match (pre_trace, diff_trace) {
+        (
+            GethTrace::Known(GethTraceFrame::PreStateTracer(PreStateFrame::Default(read))),
+            GethTrace::Known(GethTraceFrame::PreStateTracer(PreStateFrame::Diff(diff))),
+        ) => process_tx_traces(code_db, access_list, read, diff).await?,
+        _ => unreachable!(),
+    };
+
+    Ok(TxnInfo {
+        meta: tx_meta,
+        traces: tx_traces,
+    })
+}
+
+/// Fetches the transaction data for the given transaction hash.
+async fn fetch_tx_data(
+    provider: Arc<Provider<RetryClient<Http>>>,
+    tx_hash: &H256,
+) -> Result<
+    (
+        ethers::types::Transaction,
+        TransactionReceipt,
+        GethTrace,
+        GethTrace,
+    ),
+    anyhow::Error,
+> {
+    let tx_fut = provider.get_transaction(*tx_hash);
+    let tx_receipt_fut = provider.get_transaction_receipt(*tx_hash);
+    let pre_trace_fut = provider.debug_trace_transaction(*tx_hash, tracing_options());
+    let diff_trace_fut = provider.debug_trace_transaction(*tx_hash, tracing_options_diff());
+
+    let (tx, tx_receipt, pre_trace, diff_trace) =
+        futures::try_join!(tx_fut, tx_receipt_fut, pre_trace_fut, diff_trace_fut,)?;
+
+    Ok((
+        tx.ok_or_else(|| anyhow!("Transaction not found."))?,
```

Comment:
```rust
use anyhow::Context as _;
```
```suggestion
        tx.context("Transaction not found.")?,
```

Also works on `Option`s